certora/specs/SocialRecoveryModule.spec

```diff
@@ -1,26 +1,212 @@
+function sumarizeSafeIsOwner(address callee, address owner) returns bool {
+    assert callee == safeContract;
+    return safeContract.isOwner(owner);
+}
+
+// A setup function that requires Safe contract to enable the Social Recovery Module.
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
-// This is a dummy rule to verify the Safe Contract Setup with the Social Recovery
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+// Setup function that `require`s the integrity of the guardians linked list
+// in the `GuardianStorage` contract. This is needed as `invariant`s can only
+// be proven for `currentContract` and not `guardianStorageContract`, however,
+// the integrity of this list is required for rules and invariants of the
+// recovery module. For proof of this integrity, see `GuardianStorage.spec`.
+function requireGuardiansLinkedListIntegrity(address guardian) {
+    uint256 index;
+    require index < guardianStorageContract.countGuardians(safeContract);
+    require currentContract.isGuardian(safeContract, guardian) =>
+        guardianStorageContract.getGuardians(safeContract)[index] == guardian;
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+}
+
+// This integrity rule verifies that if the addGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the new guardian is added to the guardian list,
+// - and no other account (guardian or not) is affected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherAccount) {
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    currentContract.addGuardianWithThreshold(e, safeContract, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert currentContract.isGuardian(safeContract, guardian);
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount + 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be added considering ideal conditions.
+rule guardianCanAlwaysBeAdded(env e, address guardian, uint256 threshold) {
+    requireSocialRecoveryModuleEnabled();
+
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // The threshold should be greater than 0. We can't have guardian(s) with a threshold of 0.
+    require threshold > 0;
```

Comment:
Nit: Can we move this down to the other `threshold` constraint since they are related? We can even write it like:

```
    // The threshold must be greater than 0 and less than or equal to the total number of guardians.
    require 0 < threshold && threshold <= guardianStorageContract.countGuardians(safeContract);
```

Reply:
Done 👍🏾